interfaces/builtin/pipewire.go

```diff
@@ -0,0 +1,117 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"strings"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/interfaces/seccomp"
+	"github.com/snapcore/snapd/snap"
+)
+
+const pipewireSummary = `allows access to the pipewire sockets, and offer them`
+
+const pipewireBaseDeclarationSlots = `
+  pipewire:
+    allow-installation:
+      slot-snap-type:
+        - app
+        - core
+    deny-auto-connection: true
+    deny-connection:
+      on-classic: false
+`
+
+const pipewireConnectedPlugAppArmor = `
+owner /run/user/[0-9]*/ r,
+owner /run/user/[0-9]*/pipewire-[0-9] rw,
+`
+
+const pipewireConnectedPlugAppArmorCore = `
+owner /run/user/[0-9]*/###SLOT_SECURITY_TAGS###/pipewire-[0-9] rw,
+`
+
+const pipewireConnectedPlugSecComp = `
+shmctl
+`
+
+const pipewirePermanentSlotAppArmor = `
+owner @{PROC}/@{pid}/exe r,
+
+# For udev
+network netlink raw,
+/sys/devices/virtual/dmi/id/sys_vendor r,
+/sys/devices/virtual/dmi/id/bios_vendor r,
+/sys/**/sound/** r,
+
+owner /run/user/[0-9]*/pipewire-[0-9] rwk,
```

Comment:
On my system, running pipewire 1.2.7 I see:
```
srw-rw-rw-.  1 zyga zyga   0 01-24 08:29 pipewire-0
-rw-r-----.  1 zyga zyga   0 01-24 08:29 pipewire-0.lock
srw-rw-rw-.  1 zyga zyga   0 01-24 08:29 pipewire-0-manager
-rw-r-----.  1 zyga zyga   0 01-24 08:29 pipewire-0-manager.lock
```

Notably the interface does mentions neither `pipewire-[0-9]*.lock` nor the manager lock file.

Reply:
With the current version of the snap, it is writing these files to `$XDG_RUNTIME_DIR/snap.pipewire/pipewire-0*`, with the `pipewire-0` socket then symlinked to the real XDG_RUNTIME_DIR. It probably wouldn't hurt to allow the other names, but it isn't currently necessary.

For what it is worth, the lock files are not used by clients: they're just there to ensure the daemon doesn't replace the listening socket for another instance of the daemon. And the manager socket shouldn't need to be accessed outside of the snap.

Reply:
Fully agree with James. That is the reason why those were left from the interface.